crates/dbsp/src/operator/dynamic/controlled_filter.rs

```diff
@@ -0,0 +1,391 @@
+use std::borrow::Cow;
+
+use crate::{
+    algebra::{OrdZSet, ZBatch},
+    circuit::{
+        circuit_builder::RefStreamValue,
+        operator_traits::{BinaryOperator, Operator},
+    },
+    dynamic::{DataTrait, Erase},
+    trace::{Batch, BatchReader, BatchReaderFactories, Builder, Cursor},
+    Circuit, DBData, Scope, Stream, ZWeight,
+};
+
+pub struct ControlledFilterFactories<Z, E>
+where
+    Z: ZBatch,
+    E: DataTrait + ?Sized,
+{
+    batch_factories: Z::Factories,
+    errors_factory: <OrdZSet<E> as BatchReader>::Factories,
+}
+
+impl<Z, E> ControlledFilterFactories<Z, E>
+where
+    Z: ZBatch,
+    E: DataTrait + ?Sized,
+{
+    pub fn new<KType, VType, EType>() -> Self
+    where
+        KType: DBData + Erase<Z::Key>,
+        VType: DBData + Erase<Z::Val>,
+        EType: DBData + Erase<E>,
+    {
+        Self {
+            batch_factories: BatchReaderFactories::new::<KType, VType, ZWeight>(),
+            errors_factory: BatchReaderFactories::new::<EType, (), ZWeight>(),
+        }
+    }
+}
+
+impl<C, Z> Stream<C, Z>
+where
+    C: Circuit,
+    Z: ZBatch<Time = ()>,
+{
+    #[track_caller]
+    pub fn dyn_controlled_key_filter<T, E>(
+        &self,
+        factories: ControlledFilterFactories<Z, E>,
+        threshold: &Stream<C, Box<T>>,
+        filter_func: Box<dyn Fn(&T, &Z::Key) -> bool>,
+        report_func: Box<dyn Fn(&T, &Z::Key, &Z::Val, ZWeight, &mut E)>,
+    ) -> (Stream<C, Z>, Stream<C, OrdZSet<E>>)
+    where
+        T: DataTrait + ?Sized,
+        Box<T>: Clone,
+        E: DataTrait + ?Sized,
+    {
+        let error_stream_val = RefStreamValue::empty();
+        let filter = ControlledKeyFilter::new(
+            factories,
+            filter_func,
+            report_func,
+            error_stream_val.clone(),
+        );
+        let output = self.circuit().add_binary_operator(filter, self, threshold);
```

Comment:
this is a bit ugly, but I guess the ugliness is confined to this function, and this doesn't require changing add_binary_operator.
But in the future perhaps almost all operators with a closure will have an error stream?

Reply:
at least on the surface, it's pretty clean with each `Stream` method returning one or more output streams. Re fallible operators - probably yes, although I've also thought about making the error stream implicit.